### PR TITLE
[Core] Refactor `get_enabled_clouds`

### PR DESCRIPTION
**Pull request by cblmemo**
After this PR, `get_enabled_clouds` will perform a `sky.check` refresh when no public clouds are enabled.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
```bash
# start a new container and copy-paste all contents in `~/.aws/credentials` from local laptop to container
# latest master:
$ python -c 'import sky; print(sky.global_user_state.get_enabled_clouds())'
[]
# this branch:
$ python -c 'import sky; print(sky.check.get_enabled_clouds())'
AWS: Fetching availability zones mapping...[AWS]
# ^^^^^^^ some logs will be printed when calling sky.check and updating catalog
```
```bash
# ssh to sky serve controller and run:
rm -rf ~/.sky
# on local laptop, spin up an arbitrary service:
sky serve up <service-yaml>
# on latest master this will failed due to no cloud enabled; on this PR it will successfully spin up a service.
```
* Starting a new container and copy-paste all contents in `~/.aws/credentials` from local laptop to container; successfully `sky launch` a cluster without running `sky check` in advance.
- [x] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/check.py [hunk 1]
```diff
@@ -80,14 +83,51 @@ def check_one_cloud(cloud_tuple: Tuple[str, clouds.Cloud]) -> None:
     global_user_state.set_enabled_clouds(enabled_clouds)
 
 
+def _no_public_cloud_enabled_locally() -> bool:
+    locally_enabled_clouds = global_user_state.get_locally_enabled_clouds()
```
Comment (Michaelvll) [inline sky/check.py:87]: How about renaming it to `cached_enabled_clouds` and also `get_cached_enabled_clouds`?

File: sky/task.py [hunk 2]
```diff
@@ -871,8 +869,8 @@ def get_preferred_store_type(self) -> storage_lib.StoreType:
         # assert len(self.resources) == 1, self.resources
         storage_cloud = None
 
-        backend_utils.check_public_cloud_enabled()
-        enabled_storage_clouds = global_user_state.get_enabled_storage_clouds()
+        enabled_storage_clouds = storage_lib.get_enabled_storage_clouds(
+            raise_if_no_cloud_access=True)
         if not enabled_storage_clouds:
             raise ValueError('No enabled cloud for storage, run: sky check')
```
Comment (Michaelvll) [inline sky/task.py:875]: It seems the `raise_if_no_cloud_access=True` is duplicated with the `if not enabled_storage_clouds` below.

**Review by Michaelvll — COMMENTED** [review]
Thanks for submitting this issue @cblmemo! Could you write the tests that have been done in the PR description?

File: sky/resources.py [hunk 3]
```diff
@@ -5,9 +5,9 @@
```
Comment (Michaelvll) [inline sky/resources.py:5]: Should we add the ports check back that is reverted in #3175

Comment (cblmemo) [reply sky/resources.py:5]: It is still possible to cause some bugs using the current version of `get_enabled_clouds`: consider there is a public cloud available in the local database (on the controller), but the cloud does not support ports, and the user just enabled a new cloud on the local laptop; in this case the `get_enabled_clouds` will not trigger a refresh and still causing the original problem. I'm thinking to separate it to a new PR and force refresh before checking ports.

**Comment by cblmemo** [comment]
> Thanks for submitting this issue @cblmemo! Could you write the tests that have been done in the PR description?

I only run the formatted and the tests in the CI for now; lemme run smoke tests now

File: sky/resources.py (hunk 3 above)
Comment (Michaelvll) [reply sky/resources.py:5]: In that case, it will raise the error on the local machine directly, instead of only raising on the serve controller, right?

Comment (cblmemo) [reply sky/resources.py:5]: Oh that makes sense! SG

Comment (cblmemo) [reply sky/resources.py:5]: Will add it after this round of smoke test finished

**Comment by cblmemo** [comment]
> Thanks for submitting this issue @cblmemo! Could you write the tests that have been done in the PR description?

Just updated 🫡 PTAL

**Comment by Michaelvll** [comment]
> > Thanks for submitting this issue @cblmemo! Could you write the tests that have been done in the PR description?
> 
> Just updated 🫡 PTAL

Thanks for adding the tests @cblmemo! Can we also test CLIs, e.g. `sky launch` in a new env without running `sky check` in advance.

**Comment by cblmemo** [comment]
> > > Thanks for submitting this issue @cblmemo! Could you write the tests that have been done in the PR description?
> > 
> > 
> > Just updated 🫡 PTAL
> 
> Thanks for adding the tests @cblmemo! Can we also test CLIs, e.g. `sky launch` in a new env without running `sky check` in advance.



> > > Thanks for submitting this issue @cblmemo! Could you write the tests that have been done in the PR description?
> > 
> > 
> > Just updated 🫡 PTAL
> 
> Thanks for adding the tests @cblmemo! Can we also test CLIs, e.g. `sky launch` in a new env without running `sky check` in advance.

Just tested and it works well 🫡 added to the PR description. Also, all smoke tests passed too

File: sky/resources.py [hunk 4]
```diff
@@ -550,7 +550,7 @@ def _validate_and_set_region_zone(self, region: Optional[str],
             # Try to infer the cloud from region/zone, if unique. If 0 or >1
             # cloud corresponds to region/zone, errors out.
             valid_clouds = []
-            enabled_clouds = global_user_state.get_enabled_clouds()
+            enabled_clouds = sky_check.get_enabled_clouds()
```
Comment (Michaelvll) [inline sky/resources.py:553]: Should we raise if no cloud is enabled?

File: sky/resources.py [hunk 5]
```diff
@@ -661,7 +661,7 @@ def _try_validate_instance_type(self) -> None:
         else:
             # If cloud not specified
             valid_clouds = []
-            enabled_clouds = global_user_state.get_enabled_clouds()
+            enabled_clouds = sky_check.get_enabled_clouds()
```
Comment (Michaelvll) [inline sky/resources.py:664]: Should we set raise to `True` here?

File: sky/resources.py [hunk 6]
```diff
@@ -841,6 +841,21 @@ def _try_validate_ports(self) -> None:
         if self.cloud is not None:
             self.cloud.check_features_are_supported(
                 self, {clouds.CloudImplementationFeatures.OPEN_PORTS})
+        else:
+            at_least_one_cloud_supports_ports = False
+            for cloud in sky_check.get_enabled_clouds():
```
Comment (Michaelvll) [inline sky/resources.py:846]:
```suggestion
            for cloud in sky_check.get_enabled_clouds(raise_if_no_cloud_access=True):
```

**Review by Michaelvll — APPROVED** [review]
Thanks for the refactoring @cblmemo! It mostly looks good to me. Please make sure that we have thought through the cases and see if all those occurance should raise error : )

**Comment by cblmemo** [comment]
After an offline discussion, we decided to defer it after 0.5 release as it is error-prone.

**Comment by cblmemo** [comment]
@Michaelvll Just checked all cases and added raises. Re-running all smoke tests now. PTAL again 🫡

File: sky/benchmark/benchmark_utils.py [hunk 7]
```diff
@@ -167,19 +168,9 @@ def _create_benchmark_bucket() -> Tuple[str, str]:
...
-    elif 'GCP' in enabled_clouds:
-        bucket_type = data.StoreType.GCS.value
-    elif 'AZURE' in enabled_clouds:
-        raise RuntimeError(
-            'Azure Blob Storage is not supported yet. '
-            'Please enable another cloud to create a benchmark bucket.')
-    else:
-        raise RuntimeError('No cloud is enabled. '
-                           'Please enable at least one cloud.')
+    enabled_clouds = sky_check.get_enabled_clouds(raise_if_no_cloud_access=True)
```
Comment (Michaelvll) [inline sky/benchmark/benchmark_utils.py:171]: Should we use `get_enabled_storage_clouds` instead?

File: sky/check.py [hunk 8]
```diff
@@ -78,14 +81,45 @@ def check_one_cloud(cloud_tuple: Tuple[str, clouds.Cloud]) -> None:
                        f'{enabled_clouds_str}[/green]')
 
 
+def get_enabled_clouds(
```
Comment (Michaelvll) [inline sky/check.py:84]: How about renaming it to?
```suggestion
def get_cached_enabled_clouds_or_refresh(
```

File: sky/check.py [hunk 9]
```diff
@@ -78,14 +81,45 @@ def check_one_cloud(cloud_tuple: Tuple[str, clouds.Cloud]) -> None:
...
+        raise_if_no_cloud_access: bool = False) -> List[clouds.Cloud]:
+    """Returns a list of enabled clouds.
```
Comment (Michaelvll) [inline sky/check.py:86]:
```suggestion
    """Returns cached enabled clouds and if no cloud is enabled, refresh.
```

File: sky/check.py [hunk 10]
```diff
@@ -78,14 +81,45 @@ def check_one_cloud(cloud_tuple: Tuple[str, clouds.Cloud]) -> None:
...
+    """
+    if not global_user_state.get_cached_enabled_clouds():
+        try:
+            check(quiet=True)
+        except SystemExit:
+            # If no cloud is enabled, check() will raise SystemExit.
+            # Here we catch it and raise the exception later only if
+            # raise_if_no_cloud_access is set to True.
+            pass
+    cached_enabled_clouds = global_user_state.get_cached_enabled_clouds()
```
Comment (Michaelvll) [inline sky/check.py:106]: nit: one less read in normal case.
```suggestion
    cached_enabled_clouds = global_user_state.get_cached_enabled_clouds()
    if not cached_enabled_clouds:
        try:
            check(quiet=True)
        except SystemExit:
            # If no cloud is enabled, check() will raise SystemExit.
            # Here we catch it and raise the exception later only if
            # raise_if_no_cloud_access is set to True.
            pass
        cached_enabled_clouds = global_user_state.get_cached_enabled_clouds()
```

File: sky/resources.py [hunk 11]
```diff
@@ -553,7 +553,8 @@ def _validate_and_set_region_zone(self, region: Optional[str],
             # Try to infer the cloud from region/zone, if unique. If 0 or >1
             # cloud corresponds to region/zone, errors out.
             valid_clouds = []
-            enabled_clouds = global_user_state.get_enabled_clouds()
+            enabled_clouds = sky_check.get_enabled_clouds(
+                raise_if_no_cloud_access=True)
```
Comment (Michaelvll) [inline sky/resources.py:557]: Can we update the docstr of the function to include the `Raises`?

**Review by Michaelvll — COMMENTED** [review]
Thanks for the update @cblmemo! The code looks mostly good to me.

File: sky/benchmark/benchmark_utils.py (hunk 7 above)
Comment (cblmemo) [reply sky/benchmark/benchmark_utils.py:171]: Good catch! Changed. Thanks!

File: sky/check.py (hunk 8 above)
Comment (cblmemo) [reply sky/check.py:84]: Good point! Changed.

File: sky/resources.py (hunk 11 above)
Comment (cblmemo) [reply sky/resources.py]: We already have that?

image

Comment (Michaelvll) [reply sky/resources.py]: I mean the current caller function in this file :)

**Comment by cblmemo** [comment]
All smoke tests passed 🫡

File: sky/benchmark/benchmark_utils.py [hunk 12]
```diff
@@ -168,7 +168,9 @@ def _create_benchmark_bucket() -> Tuple[str, str]:
     bucket_name = f'sky-bench-{uuid.uuid4().hex[:4]}-{getpass.getuser()}'
 
     # Select the bucket type.
-    enabled_clouds = sky_check.get_enabled_clouds(raise_if_no_cloud_access=True)
+    enabled_clouds = storage_lib.get_enabled_storage_clouds(
```
Comment (Michaelvll) [inline sky/benchmark/benchmark_utils.py:171]: Should we rename this function as well: `get_cached_enabled_storage_clouds_or_refresh`?

**Review by Michaelvll — APPROVED** [review]
Thanks for this refactoring @cblmemo! LGTM.

File: sky/benchmark/benchmark_utils.py (hunk 12 above)
Comment (cblmemo) [reply sky/benchmark/benchmark_utils.py]: Good point! Renamed.

**Pull request merged**